bip-0088.mediawiki

```diff
@@ -0,0 +1,229 @@
+
+Constraints:
+
+# To avoid ambiguity, whitespace MUST NOT appear within the path template.
+# Commas within the Ranged index template MUST only appear in between index ranges.
+# To avoid ambiguity, an index range that matches a single value MUST be specified as Unit range.
+# To avoid ambiguity, an index template that matches any value from 0 to 2147483647 MUST be specified as Wildcard index template
+# For Non-unit range, range_end MUST be larger than range_start.
+# If there is more than one index range within the Ranged index template, range_start of the second and any subsequent range MUST be larger than the range_end of the preceeding range.
+# To avoid ambiguity, all representations of integer values larger than 0 MUST NOT start with character "0" (no leading zeroes allowed).
+# If hardened marker appears within any section in the path template, all preceding sections MUST also specify hardened matching.
+# To avoid ambiguity, if a hardened marker appears within any section in the path template, all preceding sections MUST also use the same hardened marker (either "h" or "'").
+# To avoid ambiguity, trailing slashes (for example, "1/2/") and duplicate slashes (for example, "0//1") MUST NOT appear in the template.
+
+It may be desireable to have fully unambiguous encoding, where for each valid path template string, there is no other valid template string that matches the exact same set of paths. This would enable someone to compare templates for equality through a simple string equality check, without any parsing.
+
+To achieve this, two extra rules are needed:
+
+* Within Ranged index template, subsequent range MUST NOT start with the value that is equal to the end of the previous range plus one. Thus, <nowiki>"{1,2,3-5}"</nowiki> is not allowed, and should be specified as <nowiki>"{1-5}"</nowiki> instead. This rule might make templates less convenient for frequent edits, though.
+
+* Only one type of hardened marker should be allowed (either "h" or "'").
+
+Instead of requiring the second extra rule, implementations can simply replace one type of marker with another in the template strings before comparing them.
+
+==Full and partial templates==
+
+If the template starts with "m/", that means that this is the "full" template, that matches the whole path.
+
+If the template does not start with "m/", that means that this is a "partial" template, and it can be used to match a part of the path, in the contexts where this might be appropriate (for example, when constraints for the suffix of the path might be dynamic, while constraints for the prefix of the path are fixed).
+
+Full template can be combined with partial template, where partial template extends full template,
+resulting in new, longer full template.
+
+Partial template can be combined with another partial template, resulting in new, longer partial template.
+
+Full template can not be combined with another full template.
```

Comment:
Can you provide examples for these?

Reply:
`m/49/0` (full)
`0/1/*` (partial)
`m/49/0/1/*` (combined)

`1/{3-5}` + `0/0/{1-5555}` => `1/{3-5}/0/0/{1-5555}` - combining two partial templates, resulting in a new partial template

That's what I meant. Do you think something like this example should go into BIP text ?

Reply:
That makes total sense to me now, maybe I was being dumb, but the examples helped. Your call.
Only question is, when combining two partial templates, how do you know the order?
Ex: 
1/{3-5} + 0/0/{1-5555} => 1/{3-5}/0/0/{1-5555}
or
1/{3-5} + 0/0/{1-5555} => 0/0/{1-5555}/1/{3-5}

Obviously, the first is more likely, but both

Reply:
When combining partial templates, the order is determined by the software that is combining it, and depend on the concrete usecase and the designation of each partial template ("account part", "suffix part" of the template, etc)

I can't even come up with a plausible example where this would be needed, but I see no need add any restriction on combining partial templates in this specification.